mosec/mixin/redis_worker.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2023 MOSEC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""MOSEC redis worker mixin.
+
+Provide another data transfer way between workers.
+
+The data will be stored in redis shared memory, while the object ID will be
+sent via the original way.
+
+    use case: large image tensors, cluster-shared data
+    benefits: more stable P99 latency
+
+"""
+
+from os import environ
+from typing import Any
+
+import redis
+
+from mosec.worker import Worker
+
+_REDIS_URL_ENV = "MOSEC_INTERNAL_REDIS_URL"
+_DEFAULT_KEY = "REDIS_SHM_IPC_KEY"
+
+
+class RedisShmIPCMixin(Worker):
+    """Redis shared memory worker mixin interface."""
+
+    # pylint: disable=no-self-use
+
+    _redis_client = None
+    _redis_key = None
+    _next_id = None
+
+    @classmethod
+    def set_redis_url(cls, url: str):
+        """Set the redis service url."""
+        environ[_REDIS_URL_ENV] = url
+
+    def _get_client(self):
+        """Get the redis client. This will create a new one if not exist."""
+        if not self._redis_client:
+            url = environ.get(_REDIS_URL_ENV)
+            if not url:
+                raise RuntimeError(
+                    "please set the redis url with "
+                    "`RedisShmIPCMixin.set_redis_url()`"
+                )
+            self._redis_client = redis.from_url(url)
+        return self._redis_client
+
+    def _get_redis_key(self):
+        """Get the redis key. This will create a new one if not exist."""
+        if not self._redis_key:
+            self._redis_key = f"{_DEFAULT_KEY}_{self._worker_id}_{self.stage}"
```

Comment:
This would make each worker a different key, so `bytes(str(client.incr(key)), encoding="utf-8")` will always start from 1? If so, the id generated will not be globally unique, hence not correct.

I think you could either:
1) use the same key across all clients
2) concatenate the key into the id (but need to further differentiate different mosec servers if they share the same redis server, since here we only use the worker id with the stage id as key, and no information about the mosec server id)

Reply:
I see, I will replace it with a universal key.